File: leet/0033-search-in-rotated-sorted-array/search-in-rotated-sorted-array.go
func search(nums []int, target int) int {
    if len(nums) == 0 {return -1}
    
    start := 0
    end := len(nums) - 1
    
    return searchWithStartEnd(nums, start, end, target);
}

func searchWithStartEnd(nums []int, start int, end int, target int) int {
    if start == end {
        if nums[start] == target {return start}
        return -1
    }
    
    if nums[start] < nums[end] {
        return binarySearch(nums, start, end, target)
    } else {
        mid := (start + end) / 2
        if nums[mid] == target {
            return mid
        } else if nums[mid] > nums[end] {
            result1 := binarySearch(nums, start, mid - 1, target)
            result2 := searchWithStartEnd(nums, mid + 1, end, target)
            
            if result1 != -1 {return result1}
            if result2 != -1 {return result2}
            return -1
        } else {
            result1 := searchWithStartEnd(nums, start, mid - 1, target)
            result2 := binarySearch(nums, mid + 1, end, target)
            
            if result1 != -1 {return result1}
            if result2 != -1 {return result2}
            return -1
        } 
    }
    
    return -1
}

func binarySearch(nums []int, start int, end int, target int) int{
    for start < end {
        mid := (start + end) / 2
        if nums[mid] == target {
            return mid
        } else if nums[mid] < target {
            start = mid + 1
        } else{
            end = mid - 1
        }
    }
    
    if nums[start] == target {return start}
    return -1
}

File: leet/0033-search-in-rotated-sorted-array/search-in-rotated-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
        if (nums.length == 0) return -1
    if (nums.length == 1) return nums[0] == target ? 0 : -1

    let idx = -1
    for(let i = 0 ; i < nums.length - 1; i++){
        if (nums[i] > nums[i + 1]) {
            idx = i
            break
        }
    }

    let find = function(start, end){
        while(start < end){
            let mid = Math.floor((start + end) / 2)
            if(nums[mid] == target){
                return mid
            }else if(nums[mid] < target){
                start = mid + 1
            }else{
                end = mid - 1
            }
        }
        return (nums[start] == target) ? start : -1
    }

    if (idx == -1){
        return find(0, nums.length - 1)
    }else{
        if (target == nums[0]){
            return 0
        }else if (target > nums[0]){
            return find(0, idx)
        }else{
            return find(idx + 1, nums.length - 1)
        }
    }
};
